Add backup transaction again

DylanDylann opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->
      
### Details
<!-- Explanation of the change or anything fishy that is going on -->
This fixes a regression introduced in [PR](https://github.com/Expensify/App/pull/35302)

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/38853
$ https://github.com/Expensify/App/issues/35173
$ https://github.com/Expensify/App/issues/38864

PROPOSAL: NA


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Section 1: 
1. Go to workspace chat
2. Create a distance request
3. Go to the request details page
4. Click Distance
5. Add a waypoint
6. Click outside the distance editor (Or click back button on mobile) to dismiss it instead of saving the edit
7. Click Distance
8. Verify that the distance editor will not show the new waypoint added in Step 5 because the edit is not saved

Section 2: 
1. Go to workspace chat.
2. Create a distance request.
3. Go to the request details page.
4. Click Distance.
5. Click the Save button without editing the addresses.
6. Verify that the map will not reload. Distance row will not show the Route pending

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same above
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same above
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/141406735/a5cb9f32-dbcb-4db8-b5e3-20f8efb9e6f0


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/141406735/f4fe21fe-7e0b-478e-9a97-2ee0114d4cfe



</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/141406735/2a84769a-82f0-496a-945f-e29ab1378bb6



</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/141406735/dd13a312-bc2f-4b45-8f81-451a448001cd



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/141406735/5f880559-84b6-4cff-a072-609ebbde86b1



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/141406735/d804db4e-0287-4e66-a8a4-00fea5d5557e



</details>

DylanDylann committed:
add backup transaction again

DylanDylann committed:
add isDraft param

DylanDylann committed:
lint fix

cubuspl42 commented on src/pages/iou/request/step/IOURequestStepDistance.js:
```diff
@@ -112,6 +114,29 @@ function IOURequestStepDistance({
+            TransactionEdit.restoreOriginalTransactionFromBackup(lodashGet(transaction, 'transactionID', ''));
+        };
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
```
@DylanDylann

How confident are you about this change? Should we consider reverting the original PR and re-evaluating the solution?

DylanDylann replied:
This change is migrated from the old page to handle backup transaction that I missed in the original PR

cubuspl42 commented:
Missing "Details"

DylanDylann commented:
@cubuspl42 I tried to revert the previous PR yesterday, but it is so late

DylanDylann commented:
@cubuspl42 We still have some problem here, I am trying to resolve

cubuspl42 commented:
I understand. You can consider marking the PR as a draft when it's being worked on actively, and some aspects are not ready.

DylanDylann committed:
merge main

DylanDylann commented:
@cubuspl42 @puneetlath Marked this PR as ready for review to get more eyes.

Before we used `transactionBackup` to handle cases that users edit distance without saving it. When going to the distance page we will create a copy of the transaction called `transactionBackup`, we still save any update to the transaction but if users don't click the save button we will restore the transaction based on `transactionBackup`.

In the [previous PR](https://github.com/Expensify/App/pull/35302), we missed to migrate the above logic. It caused these bugs https://github.com/Expensify/App/issues/38864, https://github.com/Expensify/App/issues/38853

In this PR, I moved that logic again but [this bug](https://github.com/Expensify/App/issues/38864) still be happening

https://github.com/Expensify/App/assets/141406735/ce06d288-dec1-40dc-8005-56cf81853f87

Let's see the above video, the `transactionBackup` always be empty even after creating `transactionBackup` (I suspect this is a bug of React-Native-ONYX)

This bug hasn't happened before because previously we created `transactionBackup` and used `transactionBackup` in 2 components (EditRequestDistancePage, DistanceRequest/index.tsx we can check [here](https://github.com/Expensify/App/pull/35302) for more detail). But when migrating all logics to new component IOURequestStepDistance, both creating and getting `transactionBackup` are handled in one component

cubuspl42 commented:
> I suspect this is a bug of React-Native-ONYX

Sometimes, I don't feel confident with Onyx, but if you suspect that there's a bug here, this makes this situation even more tricky 🙁

@puneetlath Do you have any hints or ideas here?

puneetlath commented:
Hmm, interesting. I don't have any theories off the top of my head. @DylanDylann what makes you think it's an Onyx bug?

DylanDylann commented:
@cubuspl42 @puneetlath  Currently we have 2 issues
1. https://github.com/Expensify/App/issues/38864
2. https://github.com/Expensify/App/issues/38853

This PR will fix https://github.com/Expensify/App/issues/38853 and I think we can process this PR to fix https://github.com/Expensify/App/issues/38853. We can handle https://github.com/Expensify/App/issues/38864 later to avoid blocking https://github.com/Expensify/App/issues/38853

DylanDylann committed:
merge main

DylanDylann committed:
merge main

DylanDylann committed:
add new field and get correct transactionID

DylanDylann commented:
@cubuspl42 After investigating again, I see this is not a bug from Onyx. I made a mistake in [here](https://github.com/Expensify/App/pull/39144#issuecomment-2031588089).

cubuspl42 commented:
@DylanDylann So what's the current state of the PR? Is everything finished and tested from your side, and things are ready for a new review round?

DylanDylann commented:
I updated this PR to fix these issues: https://github.com/Expensify/App/issues/38853 https://github.com/Expensify/App/issues/35173 https://github.com/Expensify/App/issues/38864

- For https://github.com/Expensify/App/issues/38853, I created a new field `transactionsBackup`  to save the original transaction while editing. Before we used transactionDraft, it caused bugs because we were also using transactionDraft to save information when creating distance. The `transactionBackup` has been used before as mentioned above

> Before we used transactionBackup to handle cases that users edit distance without saving it. When going to the distance page we will create a copy of the transaction called transactionBackup, we still save any update to the transaction but if users don't click the save button we will restore the transaction based on transactionBackup.

With this fix, https://github.com/Expensify/App/issues/35173 also be resolved

- For https://github.com/Expensify/App/issues/38864, I made a mistake while getting transactionBackup from Onyx. I also updated it to get transactionID from route

DylanDylann commented:
@cubuspl42 I tested and everything works well, I am uploading the video. You can start your work now.

cubuspl42 commented:
> to fix https://github.com/Expensify/App/issues/34610

I'm sorry about this mess, but please change the word "fix" to something else so we revert the auto-linking:

![image](https://github.com/Expensify/App/assets/2590174/b391eaf7-965d-4891-ac27-7e392347628e)

cubuspl42 commented:
@DylanDylann 

> Section 1:
> 
>  1. Go to workspace chat
>  2. Create a distance request
>  3. Go to the request details page
>  4. Click Distance
>  5. Add **another** waypoint

Is something missing? You say "another", but no other waypoint was mentioned before

cubuspl42 commented:
> Section 1:

Is this section platform-specific? I just tried to do this on Android and I don't think it makes sense. Please mark platform-specific testing steps.

cubuspl42 commented:
> Verify that the map will not reload.

https://github.com/Expensify/App/assets/2590174/6062b01e-1b54-419e-933e-0455bd89edd3

DylanDylann commented:
> Is this section platform-specific? I just tried to do this on Android and I don't think it makes sense. Please mark platform-specific testing steps.

Updated

DylanDylann committed:
merge main

DylanDylann commented:
@cubuspl42  We need to wait for the map to be loaded before going to the next step to prevent this bug

https://github.com/Expensify/App/assets/141406735/0c3df254-8590-487e-9976-a400608e4c31



I don't think it relates to the Distance Page. Even though we don't open the distance page, the map still reload

cubuspl42 commented:
> We need to wait for the map to be loaded before going to the next step to prevent this bug
> I don't think it relates to the Distance Page. Even though we don't open the distance page, the map still reload

Okey, thanks for confirming. Maybe we should clarify this in the QA steps, so the QA doesn't raise a false alarm? Technically speaking, the test (as phrased right now) failed in my case.

puneetlath commented on src/pages/iou/request/step/IOURequestStepDistance.js:
```diff
@@ -112,6 +114,29 @@ function IOURequestStepDistance({
+        // This effect runs when the component is mounted and unmounted. It's purpose is to be able to properly
+        // discard changes if the user cancels out of making any changes. This is accomplished by backing up the
+        // original transaction, letting the user modify the current transaction, and then if the user ever
+        // cancels out of the modal without saving changes, the original transaction is restored from the backup.
```
This comment feels kinda random in location. Would it make more sense to move it above the `useEffect`?

puneetlath reviewed (changes requested):
One very small comment.

DylanDylann committed:
update comment

DylanDylann commented:
@puneetlath Updated. All yours

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8732363773) to staging by https://github.com/puneetlath in version: 1.4.63-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

Beamanator commented on src/libs/actions/TransactionEdit.ts:
```diff
@@ -16,24 +16,24 @@ function createBackupTransaction(transaction: OnyxEntry<Transaction>) {
 
     // Use set so that it will always fully overwrite any backup transaction that could have existed before
-    Onyx.set(`${ONYXKEYS.COLLECTION.TRANSACTION_DRAFT}${transaction.transactionID}`, newTransaction);
+    Onyx.set(`${ONYXKEYS.COLLECTION.TRANSACTION_BACKUP}${transaction.transactionID}`, newTransaction);
```
This function, `createBackupTransaction`, is also used in `IOURequestStepAmount.tsx` - so i believe we should also be updating that component to subscribe to `ONYXKEYS.COLLECTION.TRANSACTION_BACKUP`

This looks like it's causing this blocker: https://github.com/Expensify/App/issues/40543

Beamanator replied:
Have a fix up here: https://github.com/Expensify/App/pull/40605

mountiny commented:
@DylanDylann Based on this comment https://github.com/Expensify/App/issues/40392#issuecomment-2067106921 the PR failed to fix https://github.com/Expensify/App/issues/38864

can you look into this? thanks

kavimuru commented:
This PR is still failing with #38864

The issue is reproducible in all the platforms

https://platform.applause.com/services/links/v1/external/57d6594cf88fd8935926a363259c65b1202c84b05f9d1736f43d4428abb14a81
https://platform.applause.com/services/links/v1/external/86476af8572e129051e9113cc9f7384ff6dbd5b82e820e69dce76e931f113995
https://platform.applause.com/services/links/v1/external/68dcf3a36100cccb2afe83c6cbfbc7e2a5ed6673eea35cfd2ef2b3eb112fed7a
https://platform.applause.com/services/links/v1/external/5b01522c2b60a20e184a4684f2803d1763d1b6bfe5f27ea6df0696e888b50baf
https://platform.applause.com/services/links/v1/external/12fd0b2061c7861b779294f36dfe061e167ac6e65ff1734289876fe55ee9e477

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8786104763) to production by https://github.com/mountiny in version: 1.4.63-21 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

cubuspl42 commented:
@DylanDylann Bump

DylanDylann commented:
@cubuspl42 What bump?

cubuspl42 commented:
Sorry then!

allroundexperts commented on src/pages/iou/request/step/IOURequestStepDistance.js:
```diff
@@ -191,6 +216,9 @@ function IOURequestStepDistance({
             setShouldShowAtLeastTwoDifferentWaypointsError(true);
             return;
         }
+        if (!isCreatingNewRequest) {
+            transactionWasSaved.current = true;
+        }
```
We should have checked if `isEditing` was `false` here as well. This caused https://github.com/Expensify/App/issues/52241
